fuel-core-interfaces/src/model/block.rs

```diff
@@ -14,136 +12,510 @@ use crate::{
+    /// the ids from the receipts of messages outputs.
+    pub fn new(
+        header: PartialFuelBlockHeader,
+        mut transactions: Vec<Transaction>,
+        message_ids: &[MessageId],
+    ) -> Self {
+        // I think this is safe as it doesn't appear that any of the reads actually mutate the data.
+        // Alternatively we can clone to be safe.
+        let transaction_ids: Vec<_> =
+            transactions.iter_mut().map(|tx| tx.to_bytes()).collect();
+        Self {
+            header: header.generate(&transaction_ids[..], message_ids),
+            transactions,
+        }
+    }
+
+    /// Get the hash of the header.
     pub fn id(&self) -> Bytes32 {
         self.header.id()
     }
 
+    /// Create a database friendly fuel block.
     pub fn to_db_block(&self) -> FuelBlockDb {
         FuelBlockDb {
             header: self.header.clone(),
             transactions: self.transactions.iter().map(|tx| tx.id()).collect(),
         }
     }
+
+    /// Get the executed transactions.
+    pub fn transactions(&self) -> &[Transaction] {
+        &self.transactions[..]
+    }
+
+    /// Get the complete header.
+    pub fn header(&self) -> &FuelBlockHeader {
+        &self.header
+    }
+
+    #[cfg(any(test, feature = "test-helpers"))]
+    pub fn transactions_mut(&mut self) -> &mut Vec<Transaction> {
+        &mut self.transactions
+    }
+
+    #[cfg(any(test, feature = "test-helpers"))]
+    pub fn header_mut(&mut self) -> &mut FuelBlockHeader {
+        &mut self.header
+    }
+}
+
+impl PartialFuelBlock {
+    pub fn new(header: PartialFuelBlockHeader, transactions: Vec<Transaction>) -> Self {
+        Self {
+            header,
+            transactions,
+        }
+    }
+
+    /// Creates a [`FuelBlockDb`] that has the
+    /// generated fields set to meaningless values.
+    ///
+    /// Hack until we figure out a better way to represent
+    /// a block in the database that hasn't been run.
+    pub fn to_partial_db_block(&self) -> FuelBlockDb {
+        FuelBlockDb {
+            header: self.header.clone().without_generated(),
+            transactions: self.transactions.iter().map(|tx| tx.id()).collect(),
+        }
+    }
+
+    /// Generate a [`FuelBlock`] after running this partial block.
+    ///
+    /// The order of the messages must be the same as they were
+    /// produced in.
+    ///
+    /// Message ids are produced by executed the transactions and collecting
+    /// the ids from the receipts of messages outputs.
+    pub fn generate(self, message_ids: &[MessageId]) -> FuelBlock {
+        FuelBlock::new(self.header, self.transactions, message_ids)
+    }
 }
 
-/// This structure is created as placeholder for future usage.
 #[derive(Clone, Debug, Default)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
-pub struct FuelBlockConsensus {}
+/// The consensus related data that doesn't live on the
+/// header.
+pub struct FuelBlockConsensus {
```

Comment:
What about making this an enum for backward compatibility as we add new consensus modes?

Reply:
Do you mean backward compatibility with production data that's living on other older fuel nodes?
We would have to make it an `#[non_exhaustive]` enum for it to be backwards compatible. Do we want to do this now for POA when we will be the only node running and can just migrate the database?

Reply:
TODO: Make trait for behavior that must be consistent across all variants of this enum

Reply:
Ok I added the enum but the trait doesn't make sense because we can't implement it in the interfaces module